### PR TITLE
feat(chat): lazy load messages

### PR DESCRIPTION
**Pull request by sdwoodbury**
### What this PR does 📖

- Only loads 10 messages for the active chat initially. then loads 10 more every 1/2 second. This allows the user to switch between conversations with almost zero lag. 
- The compose component is refactored to actually allow code maintenance. It's a big change but without this, the `rsx!` macro to render messages consists of what seems like 1000 lines of code and is nearly impossible to modify.  
- Some actions in State were refactored to require less cloning. This was accomplished by using references and replacing structs such as `Chat`, `Message`, and `Identity` with their `Uuid` or `DID` where appropriate. 

### Which issue(s) this PR fixes 🔨

- Resolve #

### Additional comments 🎤
Some variables in the friends pages were change unnecessarily. This is because I thought I could use a single reference instead of cloning the variable X times (for the `move` closures). The IDE didn't show the error at first and then I had to go change it back.

### CODE REVIEW

File: ui/src/components/chat/compose.rs [hunk 1]
```diff
@@ -448,166 +481,244 @@ fn get_messages(cx: Scope<ComposeProps>) -> Element {
...
+        active_chat_id: _,
+    } = cx.props;
+
+    let messages = &group.messages;
+    let last_message = messages.last().unwrap().message;
+    let sender = state.read().get_identity(&group.sender);
+    let sender_name = sender.username();
+    let active_language = &state.read().settings.language;
+
+    // todo: investigate why updating one's profile picture affects their status.
```
Comment (dariusc93) [inline ui/src/components/chat/compose.rs]: Is this still the case since #430 was merged?

**Review by phillsatellite — DISMISSED** [review]

File: ui/src/components/chat/compose.rs (hunk 1 above)
Comment (sdwoodbury) [reply ui/src/components/chat/compose.rs]: it doesn't look like it. I removed the comment.

File: common/src/state/action.rs [hunk 2]
```diff
@@ -135,25 +127,19 @@ pub enum Action {
     RemoveFromSidebar(Uuid),
     /// Adds or removes a chat from the favorites page
     #[display(fmt = "ToggleFavorite")]
-    ToggleFavorite(Chat),
+    ToggleFavorite(&'a Uuid),
```
Comment (dariusc93) [inline common/src/state/action.rs:130]: `Uuid` implements `Copy` so you could easily dereference it without needing to borrow it

File: common/src/state/action.rs [hunk 3]
```diff
@@ -135,25 +127,19 @@ pub enum Action {
...
     #[display(fmt = "RemoveReaction")]
     RemoveReaction(Uuid, Uuid, String),
-    /// Reply to a given message by ID
-    #[display(fmt = "Reply")]
-    Reply(Chat, Message),
-    /// Prep the UI for a message reply.
+    /// chat id, message id
     #[display(fmt = "StartReplying")]
-    StartReplying(Chat, Message),
+    StartReplying(&'a Uuid, &'a ui_adapter::Message),
```
Comment (dariusc93) [inline common/src/state/action.rs:142]: ditto for `Uuid`

**Review by dariusc93 — APPROVED** [review]
Looks good overall. I did leave a comment but isnt blocking to this PR.

**Review by Jekrimo — APPROVED** [review]

File: common/src/state/action.rs (hunk 3 above)
Comment (sdwoodbury) [reply common/src/state/action.rs:142]: thanks, I'll make these changes in the next iteration

**Pull request merged**